lib/assign.js

```diff
@@ -32,16 +32,297 @@ mysql> select action,count(*) from mdl_log where module='assign' group by action
-        Need to read some code...
+        userid  => mdl_user.id
+        course  => mdl_course.id
+        module  => mdl_module.name
+        cmid    => mdl_course_modules.id
+        url: id => cmid
+        info    => mdl_assign.id
+
+        mysql> SELECT count(*) FROM mdl_log WHERE module='assign' AND action = 'add';
+        +----------+
+        | count(*) |
+        +----------+
+        |      315 |
+        +----------+
         */
+
+        sql_old:
+            'SELECT log.*, ' +
+                'u.username AS u_username, u.email AS u_email, ' +
+                'c.shortname AS course_shortname, ' +
+                'a.name AS assign_name ' +
+                'FROM mdl_log log ' +
+                'LEFT JOIN mdl_course c ON log.course = c.id ' +
+                'LEFT JOIN mdl_course_modules cm ON log.cmid = cm.id ' +
+                'LEFT JOIN mdl_assign a ON cm.instance = a.id ' +
+                'JOIN mdl_user u ON log.userid = u.id ' +
+                "WHERE log.module = 'assign'" +
+                "AND log.action = 'add'" +
+                "AND " + restrict_clause,
+
+        sql_old_2pass: (row) => {
+            var m = row.url.match(/id=(\d+)/);
+            if (!m){
+                console.log('No course module ID in URL ' + JSON.stringify(row));
+                return null;
+            }
+        },
+
+        sql_match: (row) => {
+            return row.cm_id ?
+                mysql.format(
+                    'SELECT cm.id AS cm_id ' +
+                    'FROM mdl_course_modules cm ' +
+                    'LEFT JOIN mdl_course c ON c.id = cm.course ' +
+                    'JOIN mdl_assign a ON a.id = cm.instance ' +
+                    'JOIN mdl_modules m ON m.id = cm.module ' +
+                    'JOIN mdl_user u ON (BINARY u.email = ? OR u.username = ?) ' +
+                    'WHERE m.name = ? AND c.shortname = ? AND a.name = ?',
+                    [
+                        row["u_email"],
+                        row["u_username"],
+                        'assign',
+                        row["course_shortname"],
+                        row["assign_name"],
+                    ]
+                )
+                :
+                console.log('SELECT cm.id AS cm_id ' +
+                    'FROM mdl_course_modules cm  ' +
+                    'LEFT JOIN mdl_course c ON c.id = cm.course  ' +
+                    'JOIN mdl_assign a ON a.id = cm.instance  ' +
+                    'JOIN mdl_modules m ON m.id = cm.module  ' +
+                    "JOIN mdl_user u ON (BINARY u.email = '" + row["u_email"] + "' OR u.username = '" + row["u_username"] + "')  " +
+                    "WHERE m.name = 'assign' AND c.shortname = '" + row["course_shortname"] + "' AND a.name = '" + row["assign_name"] + "'",
+                    ' \n ',
+                    'No course module ID from DB ' + JSON.stringify(row));
+                return null;
```

Comment:
The "return null" statement here is dead code... Did you mean to have console.log()'s return value be the result of sql_match()?

Also, you'll already get the content of row[] printed out by the caller if you return null here, so you might not need console.log() at all.